src/main/java/io/numaproj/numaflow/batchmapper/Service.java

```diff
@@ -0,0 +1,126 @@
+package io.numaproj.numaflow.batchmapper;
+
+import akka.actor.ActorRef;
+import akka.actor.ActorSystem;
+import akka.actor.AllDeadLetters;
+import com.google.protobuf.Empty;
+import io.grpc.Status;
+import io.grpc.stub.StreamObserver;
+import io.numaproj.numaflow.batchmap.v1.BatchMapGrpc;
+import io.numaproj.numaflow.batchmap.v1.Batchmap;
+import lombok.AllArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+
+import java.time.Instant;
+import java.util.concurrent.CompletableFuture;
+
+
+@Slf4j
+@AllArgsConstructor
+class Service extends BatchMapGrpc.BatchMapImplBase {
+
+    private final BatchMapper batchMapper;
+    public static final ActorSystem batchMapActorSystem = ActorSystem.create("batchmap");
+
+
+    static void handleFailure(
+            CompletableFuture<Void> failedFuture,
+            StreamObserver<Batchmap.BatchMapResponse> responseStreamObserver) {
+        new Thread(() -> {
+            try {
+                failedFuture.get();
+            } catch (Exception e) {
+                e.printStackTrace();
+                var status = Status.UNKNOWN.withDescription(e.getMessage()).withCause(e);
+                responseStreamObserver.onError(status.asException());
+            }
+        }).start();
+
+    }
+
+    @Override
+    public StreamObserver<Batchmap.BatchMapRequest> batchMapFn(StreamObserver<Batchmap.BatchMapResponse> responseObserver) {
+
+        if (this.batchMapper == null) {
+            return io.grpc.stub.ServerCalls.asyncUnimplementedStreamingCall(
+                    BatchMapGrpc.getBatchMapFnMethod(),
+                    responseObserver);
+        }
+
+        DatumIteratorImpl datumStream = new DatumIteratorImpl();
+
+        CompletableFuture<Void> failedFuture = new CompletableFuture<>();
+
+        // create a shutdown actor that listens to exceptions.
+        ActorRef shutdownActorRef = batchMapActorSystem.actorOf(BatchMapShutdownActor.props(
+                failedFuture));
+
+        // subscribe for dead letters
+        batchMapActorSystem.getEventStream().subscribe(shutdownActorRef, AllDeadLetters.class);
+
+        handleFailure(failedFuture, responseObserver);
+
+        ActorRef batchMapActor = batchMapActorSystem
+                .actorOf(BatchMapActor.props(
+                        batchMapper,
+                        shutdownActorRef,
+                        responseObserver
+                ));
+
+        return new StreamObserver<Batchmap.BatchMapRequest>() {
+            @Override
+            public void onNext(Batchmap.BatchMapRequest mapRequest) {
```

Comment:
With this `onNext` and `onCompleted` combination, we collect ALL the datums to the iterator and then send the batch to the AKKA actor to iterate through. This approach doesn't leverage the power of AKKA where a single datum once received gets immediately passed to the actor to handle, like a GO channel. From a performance perspective, this is the same as passing in an array of datums.

This leads to two questions: 1. does having input as a channel help improve performance in Go SDK? and 2. if the answer to 1 is yes, then we need to properly use AKKA as we did for reduce stream, which requires two user interfaces. `processMessage` and `handleEndOfStream`, which is not as easy to use as the one in this PR where we just pass in a single iterator.

@kohlisid , @yhl25 WDYT?

Reply:
This is a valid point but the current implementation is not completely async. AFAIK even in Go the sdk gets a batch response which i feel would be the long poll. @kohlisid @yhl25 @vigith Thoughts?